peer/peerlist/list.go

```diff
@@ -0,0 +1,598 @@
+		thunks = append(thunks, t)
+		delete(pl.availablePeers, id)
+		pl.availableChooser.Remove(t, t.Subscriber())
+	}
+	return thunks
+}
+
+// removeAllUnavailablePeers will clear the unavailablePeers list and
+// return all the Peers in the list in a slice
+// Must be run in a mutex.Lock()
+func (pl *List) removeAllUnavailablePeers(toRemove map[string]*peerThunk) []*peerThunk {
+	thunks := make([]*peerThunk, 0, len(toRemove))
+	for id, t := range toRemove {
+		thunks = append(thunks, t)
+		delete(toRemove, id)
+	}
+	return thunks
+}
+
+// releaseAll will iterate through a list of peers and call release
+// on the transport
+func (pl *List) releaseAll(peers []*peerThunk) []error {
+	var errs []error
+	for _, t := range peers {
+		if err := pl.transport.ReleasePeer(t.peer, t); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+// removePeerIdentifier will go remove references to the peer identifier and release
+// it from the transport
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifier(pid peer.Identifier) error {
+	t, err := pl.removePeerIdentifierReferences(pid)
+	if err != nil {
+		// The peer has already been removed
+		return err
+	}
+
+	return pl.transport.ReleasePeer(pid, t)
+}
+
+// removePeerIdentifierReferences will search through the Available and Unavailable Peers
+// for the PeerID and remove it
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifierReferences(pid peer.Identifier) (*peerThunk, error) {
+	if t := pl.availablePeers[pid.Identifier()]; t != nil {
+		return t, pl.removeFromAvailablePeers(t)
+	}
+
+	if t, ok := pl.unavailablePeers[pid.Identifier()]; ok && t != nil {
+		pl.removeFromUnavailablePeers(t)
+		return t, nil
+	}
+
+	return nil, peer.ErrPeerRemoveNotInList(pid.Identifier())
+}
+
+// removeFromAvailablePeers remove a peer from the Available Peers list the
+// Peer should already be validated as non-nil and in the Available list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromAvailablePeers(t *peerThunk) error {
+	delete(pl.availablePeers, t.peer.Identifier())
+	pl.availableChooser.Remove(t, t.Subscriber())
+	t.SetSubscriber(nil)
+	return nil
+}
+
+// removeFromUnavailablePeers remove a peer from the Unavailable Peers list the
+// Peer should already be validated as non-nil and in the Unavailable list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromUnavailablePeers(t *peerThunk) {
+	delete(pl.unavailablePeers, t.peer.Identifier())
+}
+
+// Choose selects the next available peer in the peer list
+func (pl *List) Choose(ctx context.Context, req *transport.Request) (peer.Peer, func(error), error) {
+	if err := pl.once.WaitUntilRunning(ctx); err != nil {
+		return nil, nil, pl.newNotRunningError(err)
+	}
+
+	for {
+		p, err := pl.choose(ctx, req)
+		if p != nil {
+			t := p.(*peerThunk)
+			pl.notifyPeerAvailable()
+			t.onStart()
+			return t.peer, t.boundOnFinish, nil
+		}
+		if !yarpcerrors.IsUnavailable(err) {
```

Comment:
Ahh, I get what you were referring to in the chat yesterday.  The way we have been indicating "Failure" is by returning nil Peers. The error returned api is giving us the opening to create "hard" failures vs "maybe a peer will be added later" failures.

In the fallback for this case, if something like a shard manager wants to block until a peer for a particular shard is available it will need to have it's own "wait" functionality since the "wait" functionality here will only wait until an available peer... which should be fine I thiiiiiink.